Add CI tests for AOT Compile

priyaramani opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65441

Adding CI test to verify a simple linear model can compile fine. 
Successful run from CI logs: 

```
+ test_aot_model_compiler
+ echo 'Testing AOT model compiler'
Testing AOT model compiler
+ source test/mobile/nnc/test_aot_compile.sh
+++ python -c 'import site; print(site.getsitepackages()[0])'
++ TORCH_INSTALL_DIR=/opt/conda/lib/python3.6/site-packages/torch
++ TORCH_BIN_DIR=/opt/conda/lib/python3.6/site-packages/torch/bin
+++ dirname test/mobile/nnc/test_aot_compile.sh
++ CURRENT_DIR=test/mobile/nnc
++ MODEL=aot_test_model.pt
++ COMPILED_MODEL=aot_test_model.compiled.pt
++ COMPILED_CODE=aot_test_model.compiled.ll
++ test_aot_model_compiler
++ python test/mobile/nnc/aot_test_model.py
++ exit_code=0
++ [[ 0 != 0 ]]
++ /opt/conda/lib/python3.6/site-packages/torch/bin/test_aot_model_compiler --model aot_test_model.pt --model_name=aot_test_model --model_version=v1 --input_dims=2,2,2
The compiled model was saved to aot_test_model.compiled.pt
++ success=1
++ '[' '!' -f aot_test_model.compiled.pt ']'
++ '[' '!' -f aot_test_model.compiled.ll ']'
++ '[' -f aot_test_model.compiled.ll ']'
++ rm aot_test_model.compiled.ll
++ '[' -f aot_test_model.compiled.pt ']'
++ rm aot_test_model.compiled.pt
++ rm aot_test_model.pt
++ '[' 1 = 0 ']'
+ [[ linux-xenial-py3.6-gcc5.4-default == pytorch-linux-xenial-py3* ]]
+ assert_git_not_dirty
+ [[ linux-xenial-py3.6-gcc5.4-default != *rocm* ]]
+ [[ linux-xenial-py3.6-gcc5.4-default != *xla* ]]
++ git status --porcelain
+ git_status=
+ [[ -n '' ]]
+ test_custom_script_ops
```

Differential Revision: [D31348169](https://our.internmc.facebook.com/intern/diff/D31348169)

priyaramani committed:
Add CI tests for AOT Compile

[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile [WIP]"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile [WIP]"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile [WIP]"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile [WIP]"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile [WIP]"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile [WIP]"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile [WIP]"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile [WIP]"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile [WIP]"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile [WIP]"




[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile [WIP]"




[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31348169).

priyaramani committed:
Update on "Add CI tests for AOT Compile"


Adding CI test to verify a simple linear model can compile fine. 
Successful run from CI logs: 

```
+ test_aot_model_compiler
+ echo 'Testing AOT model compiler'
Testing AOT model compiler
+ source test/mobile/nnc/test_aot_compile.sh
+++ python -c 'import site; print(site.getsitepackages()[0])'
++ TORCH_INSTALL_DIR=/opt/conda/lib/python3.6/site-packages/torch
++ TORCH_BIN_DIR=/opt/conda/lib/python3.6/site-packages/torch/bin
+++ dirname test/mobile/nnc/test_aot_compile.sh
++ CURRENT_DIR=test/mobile/nnc
++ MODEL=aot_test_model.pt
++ COMPILED_MODEL=aot_test_model.compiled.pt
++ COMPILED_CODE=aot_test_model.compiled.ll
++ test_aot_model_compiler
++ python test/mobile/nnc/aot_test_model.py
++ exit_code=0
++ [[ 0 != 0 ]]
++ /opt/conda/lib/python3.6/site-packages/torch/bin/test_aot_model_compiler --model aot_test_model.pt --model_name=aot_test_model --model_version=v1 --input_dims=2,2,2
The compiled model was saved to aot_test_model.compiled.pt
++ success=1
++ '[' '!' -f aot_test_model.compiled.pt ']'
++ '[' '!' -f aot_test_model.compiled.ll ']'
++ '[' -f aot_test_model.compiled.ll ']'
++ rm aot_test_model.compiled.ll
++ '[' -f aot_test_model.compiled.pt ']'
++ rm aot_test_model.compiled.pt
++ rm aot_test_model.pt
++ '[' 1 = 0 ']'
+ [[ linux-xenial-py3.6-gcc5.4-default == pytorch-linux-xenial-py3* ]]
+ assert_git_not_dirty
+ [[ linux-xenial-py3.6-gcc5.4-default != *rocm* ]]
+ [[ linux-xenial-py3.6-gcc5.4-default != *xla* ]]
++ git status --porcelain
+ git_status=
+ [[ -n '' ]]
+ test_custom_script_ops
```

Differential Revision: [D31348169](https://our.internmc.facebook.com/intern/diff/D31348169)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31348169).

ljk53 commented on .jenkins/pytorch/test.sh:
```diff
@@ -273,6 +273,7 @@ test_libtorch() {
     OMP_NUM_THREADS=2 TORCH_CPP_TEST_MNIST_PATH="test/cpp/api/mnist" "$TORCH_BIN_DIR"/test_api --gtest_filter='-IMethodTest.*' --gtest_output=xml:$TEST_REPORTS_DIR/test_api.xml
     "$TORCH_BIN_DIR"/test_tensorexpr --gtest_output=xml:$TEST_REPORTS_DIR/test_tensorexpr.xml
     "$TORCH_BIN_DIR"/test_mobile_nnc --gtest_output=xml:$TEST_REPORTS_DIR/test_mobile_nnc.xml
+    test_aot_model_compiler
```
Shall we call "test_aot_model_compiler" at the bottom section where it calls "test_libtorch" instead?

ljk53 commented on .jenkins/pytorch/test.sh:
```diff
@@ -283,6 +284,12 @@ test_libtorch() {
   fi
 }
 
+test_aot_model_compiler() {
```
nit: there are already "test_tensorexpr", "test_mobile_nnc" and "test_aot_model_compiler". I think it's better to consolidate the naming.

ljk53 commented on test/mobile/nnc/CMakeLists.txt:
```diff
@@ -15,8 +15,16 @@ target_link_libraries(test_mobile_nnc PRIVATE torch gtest)
 target_include_directories(test_mobile_nnc PRIVATE ${ATen_CPU_INCLUDE})
 target_compile_definitions(test_mobile_nnc PRIVATE USE_GTEST)
 
+add_executable(test_aot_model_compiler
```
nit: shall we just name it "aot_model_compiler"? The test prefixed executables should be doing testing, while this binary itself doesn't test anything, right?

It just so happens that we build the aot compiler binary within the unit test CMakeLists file because the default libtorch CI build job doesn't build binaries, but we might change that behavior sometime in the future. So it's better not to have the rest of the test script depends on this odd fact (i.e. just calling the binary as is, rather than prefixing it with "test").

priyaramani replied:
That is right, the binary itself doesn't do the testing but it's used for testing. 
Named it with test prefix to differentiate it from the actual binary https://github.com/pytorch/pytorch/blob/master/binaries/CMakeLists.txt#L112 until we change the behavior.

ljk53 commented on test/mobile/nnc/test_aot_compile.sh:
```diff
@@ -0,0 +1,35 @@
+
+test_aot_model_compiler() {
+  python "$CURRENT_DIR"/aot_test_model.py
+  exit_code=$?
```
I think if you set "set -e -o pipefail" at the beginning of the shell script, it will automatically fail if any subshell it launches fails. You don't have to check it here explicitly.

ljk53 commented on test/mobile/nnc/test_aot_compile.sh:
```diff
@@ -0,0 +1,35 @@
+  success=1
+  if [ ! -f "$COMPILED_MODEL" ] || [ ! -f "$COMPILED_CODE" ]; then
+    echo "AOT model compiler failed to generate $COMPILED_MODEL and $COMPILED_CODE"
+    success=0
```
I think we can simply exit here. If the goal is to call "rm ..." below before exiting, I'd recommend creating a clean working directory at the beginning of the execution instead.

priyaramani committed:
Update on "Add CI tests for AOT Compile"


Adding CI test to verify a simple linear model can compile fine. 
Successful run from CI logs: 

```
+ test_aot_model_compiler
+ echo 'Testing AOT model compiler'
Testing AOT model compiler
+ source test/mobile/nnc/test_aot_compile.sh
+++ python -c 'import site; print(site.getsitepackages()[0])'
++ TORCH_INSTALL_DIR=/opt/conda/lib/python3.6/site-packages/torch
++ TORCH_BIN_DIR=/opt/conda/lib/python3.6/site-packages/torch/bin
+++ dirname test/mobile/nnc/test_aot_compile.sh
++ CURRENT_DIR=test/mobile/nnc
++ MODEL=aot_test_model.pt
++ COMPILED_MODEL=aot_test_model.compiled.pt
++ COMPILED_CODE=aot_test_model.compiled.ll
++ test_aot_model_compiler
++ python test/mobile/nnc/aot_test_model.py
++ exit_code=0
++ [[ 0 != 0 ]]
++ /opt/conda/lib/python3.6/site-packages/torch/bin/test_aot_model_compiler --model aot_test_model.pt --model_name=aot_test_model --model_version=v1 --input_dims=2,2,2
The compiled model was saved to aot_test_model.compiled.pt
++ success=1
++ '[' '!' -f aot_test_model.compiled.pt ']'
++ '[' '!' -f aot_test_model.compiled.ll ']'
++ '[' -f aot_test_model.compiled.ll ']'
++ rm aot_test_model.compiled.ll
++ '[' -f aot_test_model.compiled.pt ']'
++ rm aot_test_model.compiled.pt
++ rm aot_test_model.pt
++ '[' 1 = 0 ']'
+ [[ linux-xenial-py3.6-gcc5.4-default == pytorch-linux-xenial-py3* ]]
+ assert_git_not_dirty
+ [[ linux-xenial-py3.6-gcc5.4-default != *rocm* ]]
+ [[ linux-xenial-py3.6-gcc5.4-default != *xla* ]]
++ git status --porcelain
+ git_status=
+ [[ -n '' ]]
+ test_custom_script_ops
```

Differential Revision: [D31348169](https://our.internmc.facebook.com/intern/diff/D31348169)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31348169).

priyaramani committed:
Update on "Add CI tests for AOT Compile"


Adding CI test to verify a simple linear model can compile fine. 
Successful run from CI logs: 

```
+ test_aot_model_compiler
+ echo 'Testing AOT model compiler'
Testing AOT model compiler
+ source test/mobile/nnc/test_aot_compile.sh
+++ python -c 'import site; print(site.getsitepackages()[0])'
++ TORCH_INSTALL_DIR=/opt/conda/lib/python3.6/site-packages/torch
++ TORCH_BIN_DIR=/opt/conda/lib/python3.6/site-packages/torch/bin
+++ dirname test/mobile/nnc/test_aot_compile.sh
++ CURRENT_DIR=test/mobile/nnc
++ MODEL=aot_test_model.pt
++ COMPILED_MODEL=aot_test_model.compiled.pt
++ COMPILED_CODE=aot_test_model.compiled.ll
++ test_aot_model_compiler
++ python test/mobile/nnc/aot_test_model.py
++ exit_code=0
++ [[ 0 != 0 ]]
++ /opt/conda/lib/python3.6/site-packages/torch/bin/test_aot_model_compiler --model aot_test_model.pt --model_name=aot_test_model --model_version=v1 --input_dims=2,2,2
The compiled model was saved to aot_test_model.compiled.pt
++ success=1
++ '[' '!' -f aot_test_model.compiled.pt ']'
++ '[' '!' -f aot_test_model.compiled.ll ']'
++ '[' -f aot_test_model.compiled.ll ']'
++ rm aot_test_model.compiled.ll
++ '[' -f aot_test_model.compiled.pt ']'
++ rm aot_test_model.compiled.pt
++ rm aot_test_model.pt
++ '[' 1 = 0 ']'
+ [[ linux-xenial-py3.6-gcc5.4-default == pytorch-linux-xenial-py3* ]]
+ assert_git_not_dirty
+ [[ linux-xenial-py3.6-gcc5.4-default != *rocm* ]]
+ [[ linux-xenial-py3.6-gcc5.4-default != *xla* ]]
++ git status --porcelain
+ git_status=
+ [[ -n '' ]]
+ test_custom_script_ops
```

Differential Revision: [D31348169](https://our.internmc.facebook.com/intern/diff/D31348169)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31348169).

priyaramani committed:
Update on "Add CI tests for AOT Compile"


Adding CI test to verify a simple linear model can compile fine. 
Successful run from CI logs: 

```
+ test_aot_model_compiler
+ echo 'Testing AOT model compiler'
Testing AOT model compiler
+ source test/mobile/nnc/test_aot_compile.sh
+++ python -c 'import site; print(site.getsitepackages()[0])'
++ TORCH_INSTALL_DIR=/opt/conda/lib/python3.6/site-packages/torch
++ TORCH_BIN_DIR=/opt/conda/lib/python3.6/site-packages/torch/bin
+++ dirname test/mobile/nnc/test_aot_compile.sh
++ CURRENT_DIR=test/mobile/nnc
++ MODEL=aot_test_model.pt
++ COMPILED_MODEL=aot_test_model.compiled.pt
++ COMPILED_CODE=aot_test_model.compiled.ll
++ test_aot_model_compiler
++ python test/mobile/nnc/aot_test_model.py
++ exit_code=0
++ [[ 0 != 0 ]]
++ /opt/conda/lib/python3.6/site-packages/torch/bin/test_aot_model_compiler --model aot_test_model.pt --model_name=aot_test_model --model_version=v1 --input_dims=2,2,2
The compiled model was saved to aot_test_model.compiled.pt
++ success=1
++ '[' '!' -f aot_test_model.compiled.pt ']'
++ '[' '!' -f aot_test_model.compiled.ll ']'
++ '[' -f aot_test_model.compiled.ll ']'
++ rm aot_test_model.compiled.ll
++ '[' -f aot_test_model.compiled.pt ']'
++ rm aot_test_model.compiled.pt
++ rm aot_test_model.pt
++ '[' 1 = 0 ']'
+ [[ linux-xenial-py3.6-gcc5.4-default == pytorch-linux-xenial-py3* ]]
+ assert_git_not_dirty
+ [[ linux-xenial-py3.6-gcc5.4-default != *rocm* ]]
+ [[ linux-xenial-py3.6-gcc5.4-default != *xla* ]]
++ git status --porcelain
+ git_status=
+ [[ -n '' ]]
+ test_custom_script_ops
```

Differential Revision: [D31348169](https://our.internmc.facebook.com/intern/diff/D31348169)

[ghstack-poisoned]

priyaramani committed:
Update on "Add CI tests for AOT Compile"


Adding CI test to verify a simple linear model can compile fine. 
Successful run from CI logs: 

```
+ test_aot_model_compiler
+ echo 'Testing AOT model compiler'
Testing AOT model compiler
+ source test/mobile/nnc/test_aot_compile.sh
+++ python -c 'import site; print(site.getsitepackages()[0])'
++ TORCH_INSTALL_DIR=/opt/conda/lib/python3.6/site-packages/torch
++ TORCH_BIN_DIR=/opt/conda/lib/python3.6/site-packages/torch/bin
+++ dirname test/mobile/nnc/test_aot_compile.sh
++ CURRENT_DIR=test/mobile/nnc
++ MODEL=aot_test_model.pt
++ COMPILED_MODEL=aot_test_model.compiled.pt
++ COMPILED_CODE=aot_test_model.compiled.ll
++ test_aot_model_compiler
++ python test/mobile/nnc/aot_test_model.py
++ exit_code=0
++ [[ 0 != 0 ]]
++ /opt/conda/lib/python3.6/site-packages/torch/bin/test_aot_model_compiler --model aot_test_model.pt --model_name=aot_test_model --model_version=v1 --input_dims=2,2,2
The compiled model was saved to aot_test_model.compiled.pt
++ success=1
++ '[' '!' -f aot_test_model.compiled.pt ']'
++ '[' '!' -f aot_test_model.compiled.ll ']'
++ '[' -f aot_test_model.compiled.ll ']'
++ rm aot_test_model.compiled.ll
++ '[' -f aot_test_model.compiled.pt ']'
++ rm aot_test_model.compiled.pt
++ rm aot_test_model.pt
++ '[' 1 = 0 ']'
+ [[ linux-xenial-py3.6-gcc5.4-default == pytorch-linux-xenial-py3* ]]
+ assert_git_not_dirty
+ [[ linux-xenial-py3.6-gcc5.4-default != *rocm* ]]
+ [[ linux-xenial-py3.6-gcc5.4-default != *xla* ]]
++ git status --porcelain
+ git_status=
+ [[ -n '' ]]
+ test_custom_script_ops
```

Differential Revision: [D31348169](https://our.internmc.facebook.com/intern/diff/D31348169)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31348169).

ljk53 commented on test/mobile/nnc/test_aot_compile.sh:
```diff
@@ -0,0 +1,27 @@
+COMPILED_MODEL=aot_test_model.compiled.pt
+COMPILED_CODE=aot_test_model.compiled.ll
+
+TMP_DIR=$(mktemp -d -t build_XXX)
```
nit: build_XXX -> build_mobile_nnc

Anything starts with "build" under the pytorch src folder will be ignored by git: https://github.com/pytorch/pytorch/blob/master/.gitignore#L212

This way it should not complain about leaving local changes when the test exits.
"mktemp" will create a folder at /tmp, which won't cause local changes under pytorch src folder either, with or without the "build_" prefix.

ljk53 commented on test/mobile/nnc/test_aot_compile.sh:
```diff
@@ -0,0 +1,27 @@
+COMPILED_CODE=aot_test_model.compiled.ll
+
+TMP_DIR=$(mktemp -d -t build_XXX)
+trap 'rm -rf "$TMP_DIR"' EXIT
```
Just curious: does it fail if we don't delete the tmp folder?

priyaramani committed:
Update on "Add CI tests for AOT Compile"


Adding CI test to verify a simple linear model can compile fine. 
Successful run from CI logs: 

```
+ test_aot_model_compiler
+ echo 'Testing AOT model compiler'
Testing AOT model compiler
+ source test/mobile/nnc/test_aot_compile.sh
+++ python -c 'import site; print(site.getsitepackages()[0])'
++ TORCH_INSTALL_DIR=/opt/conda/lib/python3.6/site-packages/torch
++ TORCH_BIN_DIR=/opt/conda/lib/python3.6/site-packages/torch/bin
+++ dirname test/mobile/nnc/test_aot_compile.sh
++ CURRENT_DIR=test/mobile/nnc
++ MODEL=aot_test_model.pt
++ COMPILED_MODEL=aot_test_model.compiled.pt
++ COMPILED_CODE=aot_test_model.compiled.ll
++ test_aot_model_compiler
++ python test/mobile/nnc/aot_test_model.py
++ exit_code=0
++ [[ 0 != 0 ]]
++ /opt/conda/lib/python3.6/site-packages/torch/bin/test_aot_model_compiler --model aot_test_model.pt --model_name=aot_test_model --model_version=v1 --input_dims=2,2,2
The compiled model was saved to aot_test_model.compiled.pt
++ success=1
++ '[' '!' -f aot_test_model.compiled.pt ']'
++ '[' '!' -f aot_test_model.compiled.ll ']'
++ '[' -f aot_test_model.compiled.ll ']'
++ rm aot_test_model.compiled.ll
++ '[' -f aot_test_model.compiled.pt ']'
++ rm aot_test_model.compiled.pt
++ rm aot_test_model.pt
++ '[' 1 = 0 ']'
+ [[ linux-xenial-py3.6-gcc5.4-default == pytorch-linux-xenial-py3* ]]
+ assert_git_not_dirty
+ [[ linux-xenial-py3.6-gcc5.4-default != *rocm* ]]
+ [[ linux-xenial-py3.6-gcc5.4-default != *xla* ]]
++ git status --porcelain
+ git_status=
+ [[ -n '' ]]
+ test_custom_script_ops
```

Differential Revision: [D31348169](https://our.internmc.facebook.com/intern/diff/D31348169)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31348169).
